src/graph/transform/cuda/knn.cu

```diff
@@ -50,6 +52,192 @@ struct SharedMemory<double> {
+                                    const int64_t dim) {
+  FloatType dist = 0;
+  IdType idx = 0;
+  for (; idx < dim - 3; idx += 4) {
+    FloatType diff0 = vec1[idx] - vec2[idx];
+    FloatType diff1 = vec1[idx + 1] - vec2[idx + 1];
+    FloatType diff2 = vec1[idx + 2] - vec2[idx + 2];
+    FloatType diff3 = vec1[idx + 3] - vec2[idx + 3];
+
+    dist += diff0 * diff0 + diff1 * diff1 + diff2 * diff2 + diff3 * diff3;
+  }
+
+  for (; idx < dim; ++idx) {
+    FloatType diff = vec1[idx] - vec2[idx];
+    dist += diff * diff;
+  }
+
+  return dist;
+}
+
+/*!
+ * \brief Compute Euclidean distance between two vectors, return positive
+ *  infinite value if the intermediate distance is greater than the worst
+ *  distance.
+ */
+template <typename FloatType, typename IdType>
+__device__ FloatType euclidean_dist_with_check(const FloatType* vec1,
+                                               const FloatType* vec2,
+                                               const int64_t dim,
+                                               const FloatType worst_dist) {
+  FloatType dist = 0;
+  IdType idx = 0;
+  bool early_stop = false;
+
+  for (; idx < dim - 3; idx += 4) {
+    FloatType diff0 = vec1[idx] - vec2[idx];
+    FloatType diff1 = vec1[idx + 1] - vec2[idx + 1];
+    FloatType diff2 = vec1[idx + 2] - vec2[idx + 2];
+    FloatType diff3 = vec1[idx + 3] - vec2[idx + 3];
+
+    dist += diff0 * diff0 + diff1 * diff1 + diff2 * diff2 + diff3 * diff3;
+    if (dist > worst_dist) {
+      early_stop = true;
+      idx = dim;
+      break;
+    }
+  }
+
+  for (; idx < dim; ++idx) {
+    FloatType diff = vec1[idx] - vec2[idx];
+    dist += diff * diff;
+    if (dist > worst_dist) {
+      early_stop = true;
+      break;
+    }
+  }
+
+  if (early_stop) {
+    return std::numeric_limits<FloatType>::max();
+  } else {
+    return dist;
+  }
+}
+
+template <typename FloatType, typename IdType>
+__device__ void build_heap(IdType* indices, FloatType* dists, int size) {
+  for (int i = size / 2 - 1; i >= 0; --i) {
+    IdType idx = i;
+    while (true) {
+      IdType largest = idx;
+      IdType left = idx * 2 + 1;
+      IdType right = left + 1;
+      if (left < size && dists[left] > dists[largest]) {
+        largest = left;
+      }
+      if (right < size && dists[right] > dists[largest]) {
+        largest = right;
+      }
+      if (largest != idx) {
+        IdType tmp_idx = indices[largest];
+        indices[largest] = indices[idx];
+        indices[idx] = tmp_idx;
+
+        FloatType tmp_dist = dists[largest];
+        dists[largest] = dists[idx];
+        dists[idx] = tmp_dist;
+        idx = largest;
+      } else {
+        break;
+      }
+    }
+  }
+}
+
+template <typename FloatType, typename IdType>
+__device__ void heap_insert(IdType* indices, FloatType* dist,
+                           IdType new_idx, FloatType new_dist,
+                           int size, bool check_repeat = false) {
```

Comment:
@nv-dlasalle Are you aware of GPU priority queue implementations?  In particular this function additionally needs to ignore duplicates.

Reply:
I'm not aware of one I'd recommend.